File: frame/src/main/java/com/dreampany/frame/data/source/room/FrameDatabase.java
package com.dreampany.frame.data.source.room;

import android.content.Context;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import com.dreampany.frame.BuildConfig;
import com.dreampany.frame.data.model.State;
import com.dreampany.frame.data.model.Store;
import com.dreampany.frame.data.source.StateDao;
import com.dreampany.frame.data.source.StoreDao;
import com.dreampany.frame.misc.Constants;
import com.google.common.base.Splitter;
import com.google.common.collect.Iterables;

/**
 * Created by Hawladar Roman on 3/6/18.
 * Dreampany Ltd
 * dreampanymail@gmail.com
 */
@Database(
        entities = {
                State.class,
                Store.class
        },
        version = 8
)
public abstract class FrameDatabase extends RoomDatabase {
    private static final String DATABASE = Iterables.getLast(Splitter.on(Constants.Sep.DOT).trimResults().split(BuildConfig.APPLICATION_ID)).concat("-db");
    private static volatile FrameDatabase instance;

    synchronized public static FrameDatabase onInstance(Context context) {
        if (instance == null) {
            instance = newInstance(context, false);
        }
        return instance;
    }

    public static FrameDatabase newInstance(Context context, boolean memoryOnly) {
        RoomDatabase.Builder<FrameDatabase> builder;

        if (memoryOnly) {
            builder = Room.inMemoryDatabaseBuilder(context.getApplicationContext(), FrameDatabase.class);
        } else {
            builder = Room.databaseBuilder(context.getApplicationContext(), FrameDatabase.class, DATABASE);
        }

        return builder
                .fallbackToDestructiveMigration()
                .build();
    }

    public abstract StateDao stateDao();

    public abstract StoreDao storeDao();
}

File: frame/src/main/kotlin/com/dreampany/frame/data/source/room/FrameDatabase.kt
package com.dreampany.frame.data.source.room

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.dreampany.frame.BuildConfig
import com.dreampany.frame.data.model.State
import com.dreampany.frame.data.model.Store
import com.dreampany.frame.data.source.dao.StateDao
import com.dreampany.frame.data.source.dao.StoreDao
import com.dreampany.frame.misc.Constants

/**
 * Created by Roman-372 on 7/19/2019
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@Database(entities = [State::class, Store::class], version = 10)
abstract class FrameDatabase : RoomDatabase() {

    companion object {
        private val DATABASE = Constants.database(BuildConfig.APPLICATION_ID)
        private var instance: FrameDatabase? = null

        @Synchronized
        fun newInstance(context: Context, memoryOnly: Boolean): FrameDatabase {
            val builder: Builder<FrameDatabase>

            if (memoryOnly) {
                builder = Room.inMemoryDatabaseBuilder(context, FrameDatabase::class.java)
            } else {
                builder = Room.databaseBuilder(context, FrameDatabase::class.java,
                    FrameDatabase.DATABASE
                )
            }

            return builder
                .fallbackToDestructiveMigration()
                .build()
        }

        @Synchronized
        fun getInstance(context: Context): FrameDatabase {
            if (instance == null) {
                instance = newInstance(context, false)
            }
            return instance!!
        }
    }

    abstract fun stateDao(): StateDao

    abstract fun storeDao(): StoreDao
}